File: app/src/main/java/io/jyri/dictator/insert/InsertionText.kt
package io.jyri.dictator.insert

object InsertionText {
    fun apply(
        current: String,
        selectionStart: Int,
        selectionEnd: Int,
        insert: String,
    ): Pair<String, Int> {
        val start = selectionStart.coerceIn(0, current.length)
        val rawEnd = selectionEnd.coerceIn(0, current.length)
        val end = maxOf(start, rawEnd)
        val next = current.substring(0, start) + insert + current.substring(end)
        return next to (start + insert.length)
    }
}

File: macos/Sources/DictatorCore/InsertionText.swift
import Foundation

public enum InsertionText {
    /// Whisper often prefixes a segment with a space. Keep it as a word
    /// separator unless the insertion sits at the start of a blank prefix.
    /// Indices match Android `String` / macOS AX: UTF-16 code units.
    public static func withSeparator(prefix: String, insert: String) -> String {
        if prefix.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return String(insert.drop(while: \.isWhitespace))
        }
        return insert
    }

    public static func apply(
        current: String,
        selectionStart: Int,
        selectionEnd: Int,
        insert: String
    ) -> (text: String, cursor: Int) {
        let ns = current as NSString
        let start = clamp(selectionStart, 0, ns.length)
        let rawEnd = clamp(selectionEnd, 0, ns.length)
        let end = max(start, rawEnd)
        let prefix = ns.substring(with: NSRange(location: 0, length: start))
        let cleaned = withSeparator(prefix: prefix, insert: insert)
        let suffix = ns.substring(from: end)
        return (prefix + cleaned + suffix, start + (cleaned as NSString).length)
    }

    private static func clamp(_ value: Int, _ lower: Int, _ upper: Int) -> Int {
        min(max(value, lower), upper)
    }
}
